Tracing: Fix incorrect indentations due to reoccurring spanIDs

ivanahuckova opened the pull request:
**What this PR does / why we need it**:
This PR fixes incorrect indentation caused by the fact, that rows in the trace view didn't have unique ids. We have used only `spanId` as a unique identifier, but when old trace and new trace had span with the same `spanId`, the key for that row didn't change and react behaved unexpectedly -> wrong indentation & wrong ancestors. 

This PR fixes it by adding traceId into the key -> `${span.traceID}-${span.spanID}--${isDetail ? 'detail' : 'bar'}`. We expect that combination of `traceID` + `spanID` should be unique. 

Moreover, I have added a fix for the issue I discovered during working on this. We didn't clear detail state between the traces and therefore if we opened a detail view for span and then loaded a new trace with the same `spanId`, the detail view for that new `spanId` was opened as well.

I have also added tests for these fixes. 

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana/issues/41573

**Special notes for your reviewer**:
You can see here recording from main branch - it has 2 traces with the same spanID and in the second trace (trace2) you can see incorrect indentation.

https://user-images.githubusercontent.com/30407135/142589781-6c7a8ec3-711f-4901-a786-a33b4628413d.mov

Here is the recording from this (fixed) branch:

https://user-images.githubusercontent.com/30407135/142590006-f6561dc0-3f87-481b-93d4-77fab8374814.mov

To test this out, you can use these 2 sample traces that both have the same spanID. 

trace1:
```
{
  "batches": [
    {
      "resource": {
        "attributes": [
          {
            "key": "service.name",
            "value": {
              "stringValue": "frontend"
            }
          },
          {
            "key": "host.name",
            "value": {
              "stringValue": "frontend-1"
            }
          }
        ]
      },
      "instrumentationLibrarySpans": [
        {
          "instrumentationLibrary": {},
          "spans": [
            {
              "traceId": "AAAAAAAAAACEiiIwxeKP9A==",
              "spanId": "hIoiMMXij/Q=",
              "name": "frontend",
              "startTimeUnixNano": "1634296766948000000",
              "endTimeUnixNano": "1634296767232000000",
              "attributes": [],
              "status": {}
            }
          ]
        }
      ]
    },
    {
      "resource": {
        "attributes": [
          {
            "key": "service.name",
            "value": {
              "stringValue": "querier"
            }
          },
          {
            "key": "host.name",
            "value": {
              "stringValue": "querier-1"
            }
          }
        ]
      },
      "instrumentationLibrarySpans": [
        {
          "instrumentationLibrary": {},
          "spans": [
            {
              "traceId": "AAAAAAAAAACEiiIwxeKP9A==",
              "spanId": "MdmmKDpzjyE=",
              "parentSpanId": "hIoiMMXij/Q=",
              "name": "span from querier-1",
              "startTimeUnixNano": "1634296766990000000",
              "endTimeUnixNano": "1634296767167000000",
              "attributes": [],
              "status": {}
            }
          ]
        }
      ]
    },
    {
      "resource": {
        "attributes": [
          {
            "key": "service.name",
            "value": {
              "stringValue": "querier"
            }
          },
          {
            "key": "host.name",
            "value": {
              "stringValue": "querier-2"
            }
          }
        ]
      },
      "instrumentationLibrarySpans": [
        {
          "instrumentationLibrary": {},
          "spans": [
            {
              "traceId": "AAAAAAAAAACEiiIwxeKP9A==",
              "spanId": "mwJYmrj4CBs=",
              "parentSpanId": "hIoiMMXij/Q=",
              "name": "span from querier-2",
              "startTimeUnixNano": "1634296767133000000",
              "endTimeUnixNano": "1634296767147000000",
              "attributes": [],
              "status": {}
            }
          ]
        }
      ]
    }
  ]
}
```

trace2:
```
{
  "batches": [
    {
      "resource": {
        "attributes": [
          { "key": "service.name", "value": { "stringValue": "db" } },
          { "key": "job", "value": { "stringValue": "tns/db" } },
          { "key": "opencensus.exporterversion", "value": { "stringValue": "Jaeger-Go-2.22.1" } },
          { "key": "host.name", "value": { "stringValue": "63d16772b4a2" } },
          { "key": "ip", "value": { "stringValue": "0.0.0.0" } },
          { "key": "client-uuid", "value": { "stringValue": "39fb01637a579639" } }
        ]
      },
      "instrumentationLibrarySpans": [
        {
          "instrumentationLibrary": {},
          "spans": [
            {
              "traceId": "AAAAAAAAAABguiq7RPE+rg==",
              "spanId": "mwJYmrj4CBs=",
              "parentSpanId": "OY8PIaPbma4=",
              "name": "HTTP GET - root",
              "kind": "SPAN_KIND_SERVER",
              "startTimeUnixNano": "1627471657255809000",
              "endTimeUnixNano": "1627471657256268000",
              "attributes": [
                { "key": "http.status_code", "value": { "intValue": "200" } },
                { "key": "http.method", "value": { "stringValue": "GET" } },
                { "key": "http.url", "value": { "stringValue": "/" } },
                { "key": "component", "value": { "stringValue": "net/http" } }
              ],
              "status": {}
            }
          ]
        }
      ]
    }
  ]
}
```

ivanahuckova committed:
Create unique keys for trace rows

ivanahuckova committed:
Add tests

ivanahuckova committed:
Fix tests

connorlindsey commented on public/app/features/explore/TraceView/TraceView.test.tsx:
```diff
@@ -143,6 +143,27 @@ describe('TraceView', () => {
+        <TraceView exploreId={ExploreId.left} dataFrames={[frameNew]} splitOpenFn={() => {}} />
+      </Provider>
+    );
+    //Process is in detail view
```
nit: Process in not in view

ivanahuckova replied:
```suggestion
```

connorlindsey reviewed (approved):
Tested on your branch and compared with ops. Looks great! Nice job finding the bug with the duplicate span id's! Makes sense in retrospect why the bug was happening but must've been hard to find

ivanahuckova committed:
Update public/app/features/explore/TraceView/TraceView.test.tsx

ivanahuckova committed:
Trigger Build

ivanahuckova committed:
Met push
:rge branch 'ivana/fix-trace-leftovers' of https://github.com/grafana/grafana into ivana/fix-trace-leftovers

aocenas commented on public/app/features/explore/TraceView/TraceView.tsx:
```diff
@@ -50,7 +56,16 @@ export function TraceView(props: Props) {
+    if (frame && prevFrame && !isEqual(prevFrame, frame)) {
+      clearDetailStates();
+    }
+  }, [frame, prevFrame, clearDetailStates]);
```
I think we could just have `frame` as dependency can't we? I mean if we don't need deep equal here, then the dependency array is checked whether it changes and basically giving you the `frame != prevFrame` check out of the box.

aocenas replied:
Also maybe it could make sense to put this effect inside `useDetailState`

markoknez commented:
Not sure if this one is the culprit but reading the description I would guess that this PR produced now a bug where rows have wrong service name reported. TraceId/SpanId is not unique and it's used by zipkin to report the difference between client-send and server-receive (and opposite on the return path). You can find the mention of this in documentation - https://zipkin.io/pages/instrumenting.html. Under **When to set Span.timestamp and duration** it is mentioned 

> Zipkin merges spans together that share the same trace and span ID. The most common case of this is to merge a span reported by both the client (cs, cr) and the server (sr, ss). For example, the client starts a span, logging “cs” and propagates it via B3 headers, the server continues that span by logging “sr”.

connorlindsey commented:
@markoknez If you're experiencing this bug, could you file a bug report in this repo with steps to reproduce the issue and we can take a look? Thanks
